src/runtime/swg-client-event-manager-test.js

```diff
@@ -0,0 +1,155 @@
+
+/** @type {!EventManagerApi.SwgClientEvent} */
+const DEFAULT_EVENT = {
+  eventType: DEFAULT_TYPE,
+  eventOriginator: DEFAULT_ORIGIN,
+  isFromUserAction: null,
+  additionalParameters: {},
+};
+
+describes.sandboxed('EventManager', {}, () => {
+  it('should throw an error for invalid events', () => {
+    /** @type {!EventManagerApi.SwgClientEvent} */
+    let event = {
+      eventType: DEFAULT_TYPE,
+      eventOriginator: DEFAULT_ORIGIN,
+      isFromUserAction: null,
+      additionalParameters: {},
+    };
+    const eventMan = new SwgClientEventManager();
+
+    let errorCount = 0;
+    const tryIt = () => {
+      try {
+        eventMan.logEvent(event);
+      } catch (e) {
+        errorCount++;
+      }
+    };
+
+    tryIt();
+    expect(errorCount).to.equal(0);
+
+    //validate event type
+    event.eventType = BAD_VALUE;
+    tryIt();
+    expect(errorCount).to.equal(1);
+    event.eventType = null;
+    tryIt();
+    expect(errorCount).to.equal(2);
+    event.eventType = OTHER_TYPE;
+    tryIt();
+    expect(errorCount).to.equal(2);
+    event.eventType = DEFAULT_TYPE;
+
+    //validate event originator
+    errorCount = 0;
+    event.eventOriginator = BAD_VALUE;
+    tryIt();
+    expect(errorCount).to.equal(1);
+    event.eventOriginator = null;
+    tryIt();
+    expect(errorCount).to.equal(2);
+    event.eventOriginator = OTHER_ORIGIN;
+    tryIt();
+    expect(errorCount).to.equal(2);
+    event.eventOriginator = DEFAULT_ORIGIN;
+
+    //validate isFromUserAction
+    errorCount = 0;
+    event.isFromUserAction = BAD_VALUE;
+    tryIt();
+    expect(errorCount).to.equal(1);
+    event.isFromUserAction = true;
+    tryIt();
+    expect(errorCount).to.equal(1);
+    event.isFromUserAction = false;
+    tryIt();
+    expect(errorCount).to.equal(1);
+    event.isFromUserAction = null;
+
+    //validate additionalParameters
+    errorCount = 0;
+    event.additionalParameters = BAD_VALUE;
+    tryIt();
+    expect(errorCount).to.equal(1);
+    event.additionalParameters = null;
+    tryIt();
+    expect(errorCount).to.equal(1);
+    event.additionalParameters = {IAmValid: 5};
+    tryIt();
+    expect(errorCount).to.equal(1);
+    event.additionalParameters = {};
+
+    //validate null object
+    errorCount = 0;
+    event = null;
+    tryIt();
+    expect(errorCount).to.equal(1);
+  });
+
+  it('should be able to listen for events', function*() {
+    const eventMan = new SwgClientEventManager();
+    let receivedEventsCount = 0;
+    const callback = () => receivedEventsCount++;
+
+    //verify it can listen to 1
+    eventMan.addListener(callback);
+    yield eventMan.logEvent(DEFAULT_EVENT);
+
+    expect(receivedEventsCount).to.equal(1);
+    //verify it can listen to 2 at the same time
+    eventMan.addListener(callback);
+    yield eventMan.logEvent(DEFAULT_EVENT);
+    expect(receivedEventsCount).to.equal(3);
+  });
+
+  it('should be able to filter out some events', function*() {
```

Comment:
This is a good test. This test however does not verify if the right events were filtered. If the implementation did the opposite behavior (when stop executing is the return value, it actually continues and otherwise returns), this test will still pass without actually performing the task.

Reply:
I think it does because it log's an event that it expects to be filtered and then immediately checks the count to verify it is 0.  Then it logs an event that shouldn't be filtered and immediately checks to see the count is 1.  If it were doing the opposite of what was expected the first check (for 0) would fail.

Reply:
you are correct. it wont pass the first check. Still do you think there is value in testing that the filterer is filtering out the events that must be dropped or keeping those that should be processed?

Reply:
Yes, that's why I wrote this test.